Improve assets table display

somebody1234 opened the pull request:
### Pull Request Description
- Closes #7463
  - Makes table header sticky
  - Clips table body so it does not overlap table header

Other changes:
- Clip table header row so it does not overlap extra columns selector
- Hide extra columns selector on local backend (PM backend)
- Focus search bar if the keypress will type regular text
- Change row height from 40px to 32px
- Add "Share" button when the editor is open
- Make entire area of backend selector (Cloud <-> Local) clickable (previously, the padding was not clickable)
- Remove the up-arrow icon to open a project. Projects are now opened by switching to the project tab using the tab selector on the top left (or by double clicking the row).
- Fix opening newly created folder (previously its entries were appended to the end, rather than under the folder)
- Indent background of "name" column (the first column)
- Minor code style changes
- Add background back to "change password" modal (oops)
- Hide "open" context menu entry and show "stop" entry, when a project is currently running
  - :information_source: It might be a good idea to support the "open" action on directories as well, however it is difficult without the assets table refactor in #7540. As such, this functionality will not be added in this PR.
- Fix horizontal padding on "sign in" user menu entry
- Hide email/password validation when using oauth logins

More fixes for assets list:
- Project is inserted at start of list when there are no existing projects
- Project is inserted at start of children when there are no existing children
- Deleting a folder collapses it (hides its descendants)
- Adding children to a newly created folder puts them at the correct depth, rather than depth 1

### Screenshots and screencasts

Table header, table body clipping, and hiding extra columns selector:

[screen-recorder-sat-aug-05-2023-00-31-09.webm](https://github.com/enso-org/enso/assets/4046547/a7262619-6fe0-4786-b059-c0e8c19a41a2)

New row height, and indented background for name column:

![image](https://github.com/enso-org/enso/assets/4046547/0e9115db-418f-4b92-bb09-34a36f259738)

### Important Notes
None

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] ~~Unit tests have been written where possible.~~
  - [x] ~~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~~

somebody1234 committed:
Sticky header; hide extra column selector on local backend

somebody1234 committed:
Clip header row to remove background behind extra columns selector

somebody1234 commented:
there are still other changes that need to be made, like the scrolling name column, and indented backgrounds for the first cell.
these are (probably) a lot more difficult though, so they are not being implemented here.

somebody1234 committed:
Minor bug fix

somebody1234 committed:
Move "shared with" column's email-on-hover to the left so it is not cut off by clip-path

somebody1234 committed:
Fix width of "+" button in "shared with" column

somebody1234 committed:
Fix scrollbar appearance

somebody1234 committed:
Merge branch 'develop' into wip/sb/improve-assets-table-display

somebody1234 commented:
@indiv0 mind QAing this one?

somebody1234 committed:
Adjust row height

somebody1234 committed:
Add share button to top right when project is open

somebody1234 committed:
Make clickable area cover entire backend switcher (cloud/local) label

somebody1234 committed:
Remove up-arrow icon

The editor is now opened by switching to the editor tab using the tab selector on the top left

somebody1234 committed:
Fix opening newly created folder; change some error messages

somebody1234 committed:
Typing now focuses the search bar

somebody1234 committed:
Adjust icons; adjust column styles

somebody1234 committed:
More style fixes

somebody1234 committed:
Minor code style changes

somebody1234 committed:
Indent background of name column

somebody1234 committed:
ix row widths

somebody1234 committed:
Prevent type-anywhere-to-search from stealing focus from another input

somebody1234 committed:
Move visibility state from `AssetRow` into `TableRow`

This prevents `AssetRow`s from losing state when hidden and re-shown

somebody1234 committed:
Merge branch 'develop' into wip/sb/improve-assets-table-display

somebody1234 committed:
Fix `startup.project` again

somebody1234 committed:
Disable page switcher button for current page

somebody1234 committed:
Don't close editor if closing different project

somebody1234 committed:
Merge branch 'develop' into wip/sb/improve-assets-table-display

somebody1234 committed:
Fix editor being closed when different project is being closed

somebody1234 committed:
Clear list of expanded directories when switching backend

somebody1234 committed:
Merge branch 'develop' into wip/sb/improve-assets-table-display

somebody1234 committed:
Fix bug in dev mode

somebody1234 committed:
Merge branch 'develop' into wip/sb/improve-assets-table-display

somebody1234 commented:
i've only fixed merge conflicts here, not tested yet

somebody1234 committed:
Style fixes

somebody1234 committed:
Fix project being inserted at start of cloud drive when there are no existing projects

somebody1234 committed:
Fix projects being inserted at the start of a folder's list of children

somebody1234 committed:
Collapse a directory when it is being deleted

somebody1234 committed:
Set correct depth for new children of newly created folders

somebody1234 commented:
edit: done *some* testing

somebody1234 committed:
Merge branch 'develop' into wip/sb/improve-assets-table-display

somebody1234 committed:
Protect against JS errors thrown by IDE bundle

somebody1234 committed:
Fix context menu position on `localhost:8080`

somebody1234 committed:
Merge branch 'develop' into wip/sb/improve-assets-table-display

somebody1234 committed:
Merge branch 'develop' into wip/sb/improve-assets-table-display

somebody1234 committed:
Replace "open" context menu action with "close" action when project is running

indiv0 commented:
@somebody1234 QA failed.

```sh
./run gui watch \
    --skip-wasm-opt \
    --wasm-profile=dev \
    --skip-version-check \
    --profiling-level=objective \
    --wasm-log-level=info \
    --wasm-uncollapsed-log-level=info
```

- ✅ GUI at http://localhost:8080 opens
- ✅ Can login
- ✅ Dashboard shows up
- ✅ Nodes in cloud project are visible
- ❌ UI in cloud project looks a bit broken in FF, is this intentional?


<details>
<summary>
Screenshot
</summary>
<img width="830" alt="Pasted image 20230817000116" src="https://github.com/enso-org/enso/assets/641547/0948b246-e7d0-43a5-a96f-75cd19745c6c">
</details>

- ❌ Component browser fails to open with compile error


<details>
<summary>
Screenshot
</summary>
<img width="829" alt="Pasted image 20230817000618" src="https://github.com/enso-org/enso/assets/641547/dd37da80-4ca2-4d64-b0c8-ed70b0f0040c">
</details>

- ✅ Table header is sticky
- ✅ Table body does not overlap table header
- ❌ Extra columns selector scrolls with projects rather than remaining sticky


https://github.com/enso-org/enso/assets/641547/e997f205-7c30-4e87-a34c-3fac604614bc

- ✅ Extra columns selector hidden on local backend
- ✅ Focus search bar if the keypress will type regular text
- ❌ Add "Share" button when the editor is open
  - **Not sure what this is referring to or how to find it?**
- ✅ Make entire area of backend selector (Cloud <-> Local) clickable (previously, the padding was not clickable)
- ✅ Up-arrow icon to open a project was removed.
- ✅ Double-clicking the row opens the project.
- ✅ Opening a newly created folder appends it's entries under the folder, rather than at the end
- ✅ Background of "name" column is indented.
- ✅ "Change password" modal has background
- ✅ Hide "open" context menu entry and show "stop" entry, when a project is currently running
  - **It says "close" rather than "stop", but I'm assuming this is intentional?**
- ✅ Project is inserted at start of list when there are no existing projects
- ✅ Project is inserted at start of children when there are no existing children
- ❌ Deleting a folder collapses it (hides its descendants)
  - ❌ Not able to delete folders with items in there
  - ❌ Deleting a folder says "Unable to delete project", which is confusing


<details>
<summary>
Screenshot
</summary>
<img width="515" alt="Pasted image 20230817041248" src="https://github.com/enso-org/enso/assets/641547/be162177-6bef-435b-919e-3dd6b0c93ce2">
</details>

- ❌ Adding children to a newly created folder puts them at the correct depth, rather than depth 1
  - **MOSTLY WORKS** Except I found a weird bug where a folder shows up multiple times if created in a folder in just the right way:
    1. Create a folder (folder A)
    2. Create a subfolder (folder A.1)
    3. Try to delete folder A (fails with the above error)
    4. Create a subsubfolder (folder A.1.A)
    5. Observe that the subsubfolder appears in the list twice


<details>
<summary>
Screenshot
</summary>
<img width="503" alt="Pasted image 20230817041602" src="https://github.com/enso-org/enso/assets/641547/583a8d5b-e529-4aee-9273-c0f463bb0c9e">
</details>

somebody1234 commented:
@indiv0
> ui broken

it seems like you are using an outdated version of the IDE wasm, the traffic lights should no longer be present on browser (non-electron)

> cb issue

very weird... can you reliably repro this?

> Add "Share" button when the editor is open

oops - the green "share" button appears in the top right in the user bar, when a cloud project is opened in the IDE. it's visible in your first attached screenshot

> Deleting a folder collapses it

oops, will fix soon

somebody1234 commented:
looking at the video it also seems like the background is missing, and the column headers are scrolling slightly... i'll have to investigate

somebody1234 commented:
re: "close" rather than "stop" - yeah, it's supposed to be the opposite of "open". i probably just misremembered when editing the description of this issue

somebody1234 commented:
oh about the column headers scrolling slightly... upon closer inspection it looks like that's overscroll - i'm assuming that's fine so leaving as-is. it's an easy change either way

somebody1234 committed:
Style and text fixes

somebody1234 commented:
> Not able to delete folders with items in there

currently seems to be disallowed by the backend. i'd imagine it will be re-enabled once the trash category is implemented.

unfortunately i think it's difficult to determine whether a folder contains items, *especially* without doing a `listDirectory` to check whether it's empty. i think the ideal solution would be for the backend to return whether a directory is empty, when returning the directory as an asset? (or just re-enabling the ability to delete non-empty folders, assuming it was disabled at some point)

somebody1234 commented:
re: the last two issues - step (d) shouldn't be possible, as the children are collapsed even if deleting fails.
at least, i haven't been able to repro it, because the children are being collapsed. maybe if the backend responds quickly enough, the entry isn't hidden in the first place??

summary:
- fixed style issues
- fixed the incorrect text: "unable to delete project"
- the other issues are either limitations, or [no-repro]

somebody1234 committed:
Merge branch 'develop' into wip/sb/improve-assets-table-display

somebody1234 committed:
Merge branch 'develop' into wip/sb/improve-assets-table-display

somebody1234 committed:
Merge branch 'develop' into wip/sb/improve-assets-table-display

indiv0 commented:
# Discussion

> currently seems to be disallowed by the backend. i'd imagine it will be re-enabled once the trash category is implemented.
>
> unfortunately i think it's difficult to determine whether a folder contains items, especially without doing a listDirectory to check whether it's empty. i think the ideal solution would be for the backend to return whether a directory is empty, when returning the directory as an asset? (or just re-enabling the ability to delete non-empty folders, assuming it was disabled at some point)

IIRC this was disabled because DynamoDB makes the cascading changes needed to delete nested folders a pain. Especially compared to doing the same in SQL. This is why it was disabled. As for implementing it again, I think it'd be useful but I suppose it's up to @PabloBuchu to schedule it as a feature if we decide to add it.

Returning whether the directory is empty isn't the way to go, since that opens us up to [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) issues. Better to do it the way we do now, which is to try to delete and then ask forgiveness if the backend says it's not possible ;)

That being said, for the sake of UI friendliness we could, on the client side, disable deletion of folders that we *know* have items in them (i.e., if the user already opened that folder) to avoid a round-trip to the network before showing an error message, but that might be overkill optimistic UI.

# QA

QA failed :(

Some issues fixed, some remain, some new ones.

## Scenario 0

```sh
./run gui watch \
    --skip-wasm-opt \
    --wasm-profile=dev \
    --skip-version-check \
    --profiling-level=objective \
    --wasm-log-level=info \
    --wasm-uncollapsed-log-level=info
```

- ✅ UI in cloud project looks correct w/ updated IDE WASM

<details>
<summary>
Screenshot
</summary>
<img width="1469" alt="image" src="https://github.com/enso-org/enso/assets/641547/0d5388f9-2824-41c8-a77a-624b8011b220">
</details>

- ✅ Component browser successfully opens w/o compile error
    - Presumably fixed by updating IDE WASM
- ❌ Table header appears to scroll w/ assets instead of being pinned
    - ✅ In a funny twist, the extra columns selector stays properly pinned now ;)

https://github.com/enso-org/enso/assets/641547/ada1daac-2db4-413f-a57e-f68fca50f3f0

- ✅ Green "share" button appears in the top right in the user bar, when a cloud project is opened in the IDE

<details>
<summary>
Screenshot
</summary>
<img width="208" alt="image" src="https://github.com/enso-org/enso/assets/641547/f386a31f-7cd5-42c4-b504-7e6da6e08f10">
</details>

- ✅ Clicking the "share" button opens a "share" modal

<details>
<summary>
Screenshot
</summary>
<img width="652" alt="image" src="https://github.com/enso-org/enso/assets/641547/7874f966-0a2c-4725-92e0-3e9de3c533e0">
</details>

- ❌ Share modal loses focus each time a character is typed

https://github.com/enso-org/enso/assets/641547/1703b50c-049e-46be-9c73-187473bfe3f1

- ❌ Scrolling assets table w/ Cloud project opened results in odd GUI behaviour
    - View adjusts w/ very large latency after scroll
    - Potentially just a perf. issue due to having `TRACE_ALL` logging enabled?
    - Though the rest of the UI appears responsive, so it's weird

https://github.com/enso-org/enso/assets/641547/4c5c41f0-20d9-4cc9-919c-a68a4c49f7ea

- ❌ Deleting a folder collapses it (still)

https://github.com/enso-org/enso/assets/641547/12445ef9-4565-4560-b8e1-29c48b0feaee

- ❌ Deleting a folder still shows the odd double message "Unable to delete folder: unable to delete directory" (c.f., above screencast)
- ❌ Inconsistent terminology in error message: sometimes it says `folder` & sometimes it says `directory`. (c.f., above screencast)
    - Backend uses "directory" everywhere AFAICT
- ❌ (Minor) Seeing some pop-in while the icons for folders wait to load (c.f., above screencast).
    - Perhaps we could add a placeholder DOM element of the right size to avoid this?
    - This is especially weird in cases where we don't end up loading an icon at all, and then still re-adjust the positioning of the items (i.e., in an empty folder).
- ❌ Folder shows up multiple times if created in a folder in just the right way
    - Still reproducible. Just creating a folder under the subfolder immediately shows the bug.

https://github.com/enso-org/enso/assets/641547/0b8dd2a8-d96d-4df7-8024-ac35f2016fcc

somebody1234 commented:
re: TOCTOU, while that's true, it's also true for folders that the client *thinks* still contain items - so i think it doesn't *hurt* to return whether the folders are empty

somebody1234 committed:
Fix icon sizing; rename "directory" to "folder" in `remoteBackend.ts`

Also fix table cell CSS

somebody1234 committed:
Make table header row sticky again

somebody1234 commented:
> Share modal loses focus each time a character is typed

Introduced by #7602, oops. Will probably fix here but also in a hotfix too

> Scrolling assets table w/ Cloud project opened results in odd GUI behaviour

This may be a FF issue? (edit: it isn't. yes, it's likely a perf issue then - I didn't encounter it and don't seem to be able to repro)

On a side note: FF seems to be broken (or maybe just my FF):
(I *think* these are okay though, as (at least currently) FF is not a supported usecase)
- The background is broken because FF doesn't seem to support `rgba` background - specifically it treats any non-zero alpha as 100% alpha
- Opening a project immediately sets it to full loading status. This seems to be because FF does not support `calc` in `stroke-dasharray`
- `sticky` + `right-0` does not seem to work for the extra columns selector, even though it does work on Chrome

somebody1234 commented:
> Deleting a folder collapses it

ah, bad wording - this is the intended *new* (fixed) behavior - the idea being that if you want to delete a folder, you don't care about its descendants. i *think* it's possible to solve this by temporarily setting an item's children to `null`, and restoring it if the delete fails, but not sure whether it's desired behavior - what do you think?

somebody1234 committed:
Fix double error message when deleting assets

somebody1234 committed:
Fix expanding directory

somebody1234 committed:
Fix "manage permissions" modal losing focus

somebody1234 commented:
@indiv0 all issues should (finally) be fixed

somebody1234 committed:
Fix background on Firefox

somebody1234 committed:
Fix horizontal padding on "sign in" button

somebody1234 committed:
Hide email/password validation when using oauth logins

somebody1234 committed:
Merge branch 'develop' into wip/sb/improve-assets-table-display

somebody1234 committed:
Merge branch 'develop' into wip/sb/improve-assets-table-display

somebody1234 committed:
Merge branch 'develop' into wip/sb/improve-assets-table-display

indiv0 commented:
QA passed (1 minor visual hiccup aside)

- ✅ Table header no longer appears to scroll w/ assets instead of being pinned 
- ✅ Share modal no longer loses focus each time a character is typed
- ⚠️ Scrolling assets table w/ Cloud project opened results in odd GUI behaviour 
    - Still present & reproducible on Chrome, but less noticeable without a project opened
    - This is a minor visual glitch so I won't block the QA on this

https://github.com/enso-org/enso/assets/641547/e166bea2-cace-4b08-a7fc-b0240a21a3ea

- ✅ Deleting a folder collapses it (still)
- ✅ Deleting a folder no longer shows the odd double message "Unable to delete folder: unable to delete directory"
- ✅ No more inconsistent terminology in error message: now it only says "folder"
- ✅ No more pop-in while waiting for directory assets to load
- ✅ Folder no longer shows up multiple times if created in a folder in just the right way

indiv0 commented on app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/column.tsx:
```diff
@@ -216,10 +191,13 @@ function SharedWithColumn(props: AssetColumnProps) {
         >
             {(item.permissions ?? []).map(user => (
-                <UserPermissionDisplay key={user.user.user_email} user={user} />
+                <PermissionDisplay key={user.user.pk} action={user.permission}>
```
Oh no we return `pk` from the database? That's a shame. `pk` is meant to be an internal-only field name that we replace with a more friendly identifier describing what this type is. Oh well, suppose it's okay for now.

indiv0 commented on app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/assetContextMenu.tsx:
```diff
@@ -68,6 +68,9 @@ export default function AssetContextMenu(props: AssetContextMenuProps) {
         self?.permission === backendModule.PermissionAction.admin
+    const isRunningProject =
+        asset.type === backendModule.AssetType.project &&
+        backendModule.IS_PROJECT_STATE_OPENING_OR_OPENED[asset.projectState.type]
```
Strictly speaking, there's no `OPENING` state, there's `OPEN_IN_PROGRESS` and `PROVISIONED`. I do realize changing it would make this variable name very verbose, but might be worth it for consistency.

somebody1234 replied:
hmm... the intent was for it to be relatively short but still clear. it's true that renaming it as you suggested might make it consistent, but then the name just becomes a list of states for which it returns true, which i'm not sure is a good idea.

unfortunately i don't have any better ideas though

somebody1234 replied:
maybe something like `IS_PROJECT_STATE_RUNNING` or `IS_PROJECT_STATE_NOT_STOPPED`...
`IS_PROJECT_STATE_NOT_CLOSED` is also an option but not sure that's particularly clear as there are three different states which indicate that a project is not running

indiv0 replied:
Yeah I suppose leave it as-is.

indiv0 commented on app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/assetsTable.tsx:
```diff
@@ -696,17 +767,20 @@ export default function AssetsTable(props: AssetsTableProps) {
-    }, [rawDoCloseIde, /* should never change */ dispatchAssetEvent])
+    const doCloseIde = React.useCallback(
+        (project: backendModule.ProjectAsset) => {
+            // FIXME: should not do this if the project is different
```
Per the style guide (located in one of our other repos) FIXMEs should always be in the following format:
```
// FIXME [sb]: https://github.com/enso-org/enso/issues/9999
// Should not do this if the project is different.
```
That is, they should be attached to an issue so we can keep track of them.

somebody1234 replied:
oh i think this is something i meant to fix but just kinda forgot about it. i probably should just fix it instead

indiv0 commented on app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/button.tsx:
```diff
@@ -29,7 +29,7 @@ export default function Button(props: ButtonProps) {
         <SvgMask
             src={image}
-            {...(disabled && error != null ? { title: error } : {})}
+            {...(!active && disabled && error != null ? { title: error } : {})}
```
Some docs explaining the difference between `active` and `disable` on the props struct definition would be helpful. As it is, it's unclear that the two serve a different purpose.

indiv0 commented on app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/driveBar.tsx:
```diff
@@ -69,27 +69,26 @@ export default function DriveBar(props: DriveBarProps) {
+                            error="Not implemented yet."
+                            disabledOpacityClassName="opacity-20"
+                            onClick={() => {
+                                // No backend support yet.
```
Should be associated with an issue perhaps?

somebody1234 replied:
i *think* it's fine to not open an issue for this? if it's something that should be implemented on the backend then i think the backend parts will be planned separately? because there are quite a number of features that do currently lack backend support, so i'm just not sure it's a good idea to file everything as issues

indiv0 reviewed (approved):
LGTM, minus some minor changes.

somebody1234 committed:
Address review

somebody1234 committed:
Address FIXME

somebody1234 commented:
@indiv0 review issues should be fixed. re: the minor visual glitch... not sure what's happening there tbh. it uses a clip path which is updated in js so it's possible that... maybe the IDE causes stutters? but it sounds like it's happening even without an IDE open.

maybe it's OS specific? (maybe scroll events don't fire at least once every frame on macos?)
or maybe it needs pretty beefy hardware to run at 60fps (?!)

indiv0 commented:
> @indiv0 review issues should be fixed. re: the minor visual glitch... not sure what's happening there tbh. it uses a clip path which is updated in js so it's possible that... maybe the IDE causes stutters? but it sounds like it's happening even without an IDE open.
> 
> maybe it's OS specific? (maybe scroll events don't fire at least once every frame on macos?) or maybe it needs pretty beefy hardware to run at 60fps (?!)

Possibly hardware issue (this is on M2 Air 16 GB), possibly due to the 1 frame event lag in EnsoGL (though this is in React so 🤷). Yeah it's not the *worst* thing so I won't block the PR on this if you can't reproduce.

somebody1234 commented:
@indiv0 yeah idk. i think it *is* worth investigating at some point so it might be worth filing an issue (or i could file one) - it'd probably not be particularly high priority though
